src/CRM/BaseMachine.hs

```diff
@@ -101,51 +124,73 @@ data InitialState (state :: vertex -> Type) where
 -- where the transition from `initialVertex` to `finalVertex` is allowed by the machine `topology`
 data
   ActionResult
+    m
     (topology :: Topology vertex)
     (state :: vertex -> Type)
     (initialVertex :: vertex)
     (output :: Type)
   where
   ActionResult
     :: AllowedTransition topology initialVertex finalVertex
-    => state finalVertex
-    -> output
-    -> ActionResult topology state initialVertex output
+    => m (output, state finalVertex)
+    -> ActionResult m topology state initialVertex output
+
+hoistResult
+  :: (forall x. m x -> n x)
+  -> ActionResult m topology state initialVertex output
+  -> ActionResult n topology state initialVertex output
+hoistResult f (ActionResult outputStatePair) = ActionResult $ f outputStatePair
 
-instance Functor (ActionResult topology state initialVertex) where
+instance Functor m => Functor (ActionResult m topology state initialVertex) where
   fmap
     :: (a -> b)
-    -> ActionResult topology state initialVertex a
-    -> ActionResult topology state initialVertex b
-  fmap f (ActionResult state output) =
-    ActionResult state (f output)
+    -> ActionResult m topology state initialVertex a
+    -> ActionResult m topology state initialVertex b
+  fmap f (ActionResult outputStatePair) =
+    ActionResult $ first f <$> outputStatePair
+
+pureResult
+  :: (Applicative m, AllowedTransition topology initialVertex finalVertex)
+  => output
+  -> state finalVertex
+  -> ActionResult m topology state initialVertex output
+pureResult output state = ActionResult . pure $ (output, state)
+
+sequence
+  :: ActionResult Identity topology state initialVertex [output]
+  -> ActionResult [] topology state initialVertex output
+sequence (ActionResult (Identity (outputs, state))) =
+  ActionResult $ (,state) <$> outputs
```

Comment:
this is actually not used, so I'll just remove it.

Anyway, I call it `sequence` because what it does is commuting two effects, like `sequence` does, but the type signature is not the same and actually not compatible

Reply:
maybe this could actually be useful, so I'll keep it instead...

It could be generalised to

```haskell
sequence
  :: Functor f
  => ActionResult Identity topology state initialVertex (f output)
  -> ActionResult f topology state initialVertex output
```

On the other hand, `sequenceA` from `Traversable` would need to be

```haskell
sequence
  :: Functor f
  => ActionResult m topology state initialVertex (f output)
  -> f (ActionResult m topology state initialVertex output)
```

I'll add a comment to explain this